### PR TITLE
Implement distributed locking using Redlock-like algorithm

### PR DESCRIPTION
**Pull request by patrislav**

### CODE REVIEW

File: compose.go [hunk 1]
```diff
@@ -184,3 +184,10 @@ func (cs *ComposeStore[V]) ClearAll(ctx context.Context) error {
 	}
 	return nil
 }
+
+func (cs *ComposeStore[V]) GetOrSetWithLock(
+	ctx context.Context, key string, getter func(context.Context, string) (V, error),
+) (V, error) {
+	var out V
+	return out, fmt.Errorf("cachestore: GetOrSetWithLock is unsupported on ComposeStore")
+}
```
Comment (patrislav) [inline compose.go:193]: Not 100% sure what the behaviour should be in the case of ComposeStore... the simplest (and probably most sane?) option could be to only use the last store (as the last one is usually slowest but most reliable, e.g. most likely Redis)...

File: cachestore.go [hunk 2]
```diff
@@ -47,6 +47,11 @@ type Store[V any] interface {
 	// ClearAll removes all data from the cache. Only use this during debugging,
 	// or testing, and never in practice.
 	ClearAll(ctx context.Context) error
+
+	// GetOrSetWithLock returns a stored value if it exists. If it doesn't, it acquires
+	// a lock and call the getter callback to retrieve a new value. Then it stores that
+	// value in the cache and releases the lock.
+	GetOrSetWithLock(ctx context.Context, key string, getter func(context.Context, string) (V, error)) (V, error)
```
Comment (pkieltyka) [inline cachestore.go:54]: can we add `ttl time.Duration` as well..? perhaps as optional.. so..

`GetOrSetWithLock(ctx context.Context, key string, getter func(context.Context, string) (V, error), ttl ...time.Duration) (V, error)`

File: memlru/memlru.go [hunk 3]
```diff
@@ -186,6 +186,30 @@ func (m *MemLRU[V]) ClearAll(ctx context.Context) error {
 	return nil
 }
 
+func (m *MemLRU[V]) GetOrSetWithLock(ctx context.Context, key string, getter func(context.Context, string) (V, error)) (V, error) {
+	var out V
+
+	m.mu.Lock()
```
Comment (pkieltyka) [inline memlru/memlru.go:192]: hmm, technically the lock should just be on the actual key..?

File: redis/redis.go [hunk 4]
```diff
@@ -283,6 +287,55 @@ func (c *RedisStore[V]) ClearAll(ctx context.Context) error {
...
+
+	mu, err := c.newMutex(ctx, key)
+	if err != nil {
+		return out, fmt.Errorf("cachestore/redis: creating mutex failed: %w", err)
+	}
+	defer mu.Unlock()
+
+	for i := 0; ; i++ {
+		// If there's a value in the cache, return it immediately
+		out, found, err := c.Get(ctx, key)
```
Comment (pkieltyka) [inline redis/redis.go:303]: I wonder if there is a race here..? and if we should acquire the lock before the get..?

Comment (patrislav) [reply redis/redis.go:303]: Two processes might be able to both `Get` but if it's empty then only one of them will be able to acquire the lock for writing (with `TryLock`), the other one will have to wait. Then, if the first process has already finished and set the value, the second one will call `Get` again to find a value there and exit the loop. So I don't think there's a race condition here...

If we acquire the lock before the Get then the mutex would protect the reads as well, so there'd only be a single concurrent reader

**Pull request merged**